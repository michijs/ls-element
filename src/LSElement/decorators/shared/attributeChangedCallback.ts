import { LSCustomElement } from '../../types';
import { getAttributeValue } from '../../utils/getAttribute';

export function attributeChangedCallback(self: LSCustomElement, name: string, oldValue, newValue) {
  if (newValue != oldValue) {
    self.componentWillReceiveAttribute?.(name, newValue, oldValue);
    self[name] = getAttributeValue(newValue);
  }
}