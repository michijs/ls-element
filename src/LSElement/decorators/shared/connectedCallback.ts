import type { LSCustomElement } from '../../types';
import { addReflectedAttributes } from '../../properties/addReflectedAttributes';
import { executeFirstRender } from '../../DOM/executeFirstRender';

export function connectedCallback(self: LSCustomElement) {//Children and attributes should be managed after constructor
  addReflectedAttributes(self);
  if (!self.ls.alreadyRendered) {
    self.componentWillMount?.();
    executeFirstRender(self);
    self.ls.alreadyRendered = true;
    self.componentDidMount?.();
  }
}