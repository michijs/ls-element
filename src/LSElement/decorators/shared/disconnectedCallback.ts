import { removeStoredAttributes } from '../../properties/removeStoredAttributes';
import type { LSCustomElement } from '../../types';

export function disconnectedCallback(self: LSCustomElement) {
  if (self.parentNode === null) {//TODO: search a better way to validate if element does not exists anymore
    removeStoredAttributes(self);
    self.componentDidUnmount?.();
  }
}