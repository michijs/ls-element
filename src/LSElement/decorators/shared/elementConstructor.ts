import type { LSCustomElement, ShadowOption, OtherShadowOptions } from '../../types';
import { addStores } from '../../properties/addStores';
import { initStore } from '../../properties/initStore';
import { addAttributes } from '../../properties/addAttributes';
import { addStoredAttributes } from '../../properties/addStoredAttributes';
import { addObservers } from '../../properties/addObservers';

export function elementConstructor(self: LSCustomElement, shadow: ShadowOption, otherShadowOptions: OtherShadowOptions) {
  if (shadow) {
    const shadowMode = shadow;
    const attachedShadow = self.attachShadow({ mode: shadowMode, ...otherShadowOptions });
    if (shadowMode === 'closed') {
      self._shadowRoot = attachedShadow;
    }
  }
  self.ls = self.ls || {};
  initStore(self);
  addStores(self);
  addAttributes(self);
  addStoredAttributes(self);
  addObservers(self);
}