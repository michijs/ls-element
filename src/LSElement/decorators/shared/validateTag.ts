export const validateTag = (tag: string) => {
  if (tag.indexOf('-') <= 0) {
    throw new Error('You need at least 1 dash in the custom element name!');
  }
};
// TODO: ver si vale la pena