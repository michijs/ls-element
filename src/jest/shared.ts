import { MainElement } from './MainElement';

export type Result =
    'create1000Rows'
    | 'replaceAllRows'
    | 'partialUpdate'
    | 'selectRow'
    | 'swapRows'
    | 'removeRow'
    | 'createManyRows'
    | 'appendRowsToLargeTable'
    | 'clearRows'

const create1000Rows = () => {
  document.getElementById('run').click();
};
const add1000Rows = () => {
  document.getElementById('add').click();
};
const create10000Rows = () => {
  document.getElementById('runlots').click();
};
const updateEvery10Rows = () => {
  document.getElementById('update').click();
};
const swapRows = () => {
  document.getElementById('swaprows').click();
};
const getTableBody = () => {
  return Array.from(document.getElementById('tbody').children);
};
const clear = () => {
  document.getElementById('clear').click();
};
const getRowId = (element: Element) => {
  return Number(element.childNodes.item(0).childNodes.item(0).textContent);
};

export function makePerformanceTests(dataController: () => MainElement, selectByIndex: boolean) {
  const results = new Map<Result, number>();
  const saveResult = (key: Result, functionToMeasure: () => void) => {
    const t0 = performance.now();
    functionToMeasure();
    const t1 = performance.now();
    results.set(key, Number(Number(t1 - t0).toFixed(2)));
  };
  it('creates 1000 rows when clicking run', () => {
    saveResult('create1000Rows', create1000Rows);
    expect(getTableBody().length).toEqual(1000);
  });
  it('replaces 1000 rows when clicking run', () => {
    create1000Rows();
    saveResult('replaceAllRows', create1000Rows);
    expect(getTableBody().length).toEqual(1000);
    getTableBody().forEach(element => {
      expect(getRowId(element)).toBeGreaterThan(1000);
    });
  });
  it('update every 10th row 1000 rows on a table with 1000 rows when clicking update', () => {
    create1000Rows();
    saveResult('partialUpdate', updateEvery10Rows);
    expect(getTableBody().length).toEqual(1000);
    getTableBody().forEach((element, index) => {
      if (index % 10 === 0) {
        expect(element.innerHTML.includes('!!!')).toBeTruthy();
      } else {
        expect(element.innerHTML.includes('!!!')).toBeFalsy();
      }
    });
  });
  it('select a row (1000 rows)', () => {
    create1000Rows();
    saveResult('selectRow', () => dataController().select(999));
    expect(getTableBody()[selectByIndex ? 999 : 998].className).toEqual('danger');
  });
  it('swap a row (1000 rows)', () => {
    create1000Rows();
    saveResult('swapRows', swapRows);
    expect(getRowId(getTableBody()[1])).toEqual(999);
    expect(getRowId(getTableBody()[998])).toEqual(2);
  });
  it('remove a row (1000 rows)', () => {
    create1000Rows();
    saveResult('removeRow', () => dataController().delete(999));
    expect(getTableBody().length).toEqual(999);
  });
  it('creates 10000 rows when clicking runlots', () => {
    saveResult('createManyRows', create10000Rows);
    expect(getTableBody().length).toEqual(10000);
  });
  it('append 1000 rows on a large table', () => {
    create10000Rows();
    saveResult('appendRowsToLargeTable', add1000Rows);
    expect(getTableBody().length).toEqual(11000);
  });
  it('clear rows', () => {
    create1000Rows();
    saveResult('clearRows', clear);
    expect(getTableBody().length).toEqual(0);
  });
  return results;
}