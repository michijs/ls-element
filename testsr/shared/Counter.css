:host {
  display: flex;
  flex-direction: row;
}

span {
  min-width: 60px;
  text-align: center;
}
