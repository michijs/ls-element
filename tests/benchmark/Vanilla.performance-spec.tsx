import { type Browser, chromium, type Page } from "playwright-core";
import { makePerformanceTests } from "./shared";
import { describe, expect, beforeEach, afterAll, beforeAll } from "bun:test";
import { spawn } from "child_process";
const serverProcess = spawn("bun", ["run", "start"], {
  stdio: "inherit",
  env: { ...process.env, NODE_ENV: "TESTING_VANILLA" },
});

describe("Performance tests - vanilla-js", () => {
  let browser: Browser;
  let page: Page;
  beforeAll(async () => {
    browser = await chromium.launch({
      headless: true,
    });
  });
  beforeEach(async () => {
    page = await browser.newPage();
    await page.goto("http://localhost:3001", {
      waitUntil: "domcontentloaded",
    });
  });
  const results = makePerformanceTests(() => page);
  afterAll(async () => {
    expect(await results).toMatchSnapshot("Vanilla JS");
    serverProcess.kill();
    browser.close();
  });
});
