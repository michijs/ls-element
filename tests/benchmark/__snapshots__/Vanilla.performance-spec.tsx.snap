// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance tests - vanilla-js clear rows: Vanilla JS 1`] = `
Map {
  "create1000Rows" => 367.21,
  "replaceAllRows" => 276.42,
  "partialUpdate" => 178.34,
  "selectRow" => 548.01,
  "swapRows" => 161.78,
  "removeRow" => 480.54,
  "createManyRows" => 1027.21,
  "appendRowsToLargeTable" => 149.67,
  "clearRows" => 147.37,
}
`;
